### DISCUSSION TITLE
tech.tablesaw.columns.strings.NoKeysAvailableException

### DISCUSSION THREAD

**Discussion by SMVitality** [jtablesaw/tablesaw#1159] [General]
Our monitoring system is showing that tech.tablesaw.columns.strings.NoKeysAvailableException are being thrown during our application processing with the Message:
**String column can only contain 255 unique values. Column has more.**

After a bit of digging I can see that the 
public StringColumn append(String value) method is handling this exception so it doesn't bubble back up into our own code
ie -- **we never knew this was happening** 

My Application is reading in a CSV file containing ~30K rows of data
My understanding was that Tablesaw would easily deal with this sort of size of Table

??Have I mis-understood something fundamental

Thanks

**Comment by lwhite1** [comment]
The string column is dictionary encoded. It starts with a byte encoding,
and then when you have more than 255 entries, it should switch to shorts,
and then if that's not enough, it should re-encode using ints. You should
be good up to Integer.MAX_INT number of unique entries. If it's not doing
that, there is a bug. I would need to see a stack trace and code to debug.

**Reply by lwhite1** [reply]
To be clear, is the application crashing or just handling the exception and continuing to work?

**Reply by SMVitality** [reply]
Application continues (not crashing)

The concern is that we are fooling ourselves  -ie we have ONLY processed 255 rows out of ~30K

**Comment by SMVitality** [comment]
??Is the following sufficient


**Exception:**
tech.tablesaw.columns.strings.NoKeysAvailableException
**Message:**
String column can only contain 255 unique values. Column has more.
**Stacktrace:**
tech.tablesaw.columns.strings.ByteDictionaryMap.getValueId(ByteDictionaryMap.java:259)
tech.tablesaw.columns.strings.ByteDictionaryMap.append(ByteDictionaryMap.java:246)
tech.tablesaw.api.StringColumn.append(StringColumn.java:457)
com.demo.myApp.lambda$
com.demo.myApp$$Lambda$.accept
java.util.Iterator.forEachRemaining(Iterator.java:116)
java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
com.demo.myApp.setRrRisks(MyCalculator.java:945)
com.demo.myApp.calculateLine(MyCalculator.java:298)
com.demo.myApp.lambda$
com.demo.myApp$$Lambda$.get
java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1582)
java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)


Thanks

**Reply by SMVitality** [reply]
FYI
<dependency>
	<groupId>tech.tablesaw</groupId>
	<artifactId>tablesaw-core</artifactId>
	<version>0.42.0</version>
</dependency>

**Comment by lwhite1** [comment]
I don’t think this is a bug.

 Perhaps you can set your monitoring to ignore handled exceptions inside
tablesaw?  They are not meant for external consumption.

**Comment by SMVitality** [comment]
OK thanks for taking a look
So can you confirm that it is operating over >255 rows within the Column object

**Comment by lwhite1** [comment]
It always has. You should be able to see it in your own data.
